### PR TITLE
E2e token transfer

### LINKED ISSUE #165: Add E2E Tests for Token bridging
Add e2e test that bridges HTS token to Ethereum.

The test is blocked by mirror node support for querying token transfers

### PR DESCRIPTION
**Pull request by Holymir**
**Detailed description**:

**Which issue(s) this PR fixes**:
Fixes #165 

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [ ] Tests updated

### CODE REVIEW

File: e2e/e2e_test.go [hunk 1]
```diff
@@ -94,6 +96,22 @@ func Test_E2E_Only_Address_Memo(t *testing.T) {
...
 	verifyTopicMessages(setupEnv, transactionResponse, 0, t)
+	//TODO: submit the tx send amount
+}
+
+func Test_E2E_Token_Transfer(t *testing.T) {
+	setupEnv := setup.Load()
+
+	memo := fmt.Sprintf("%s-0-0", receiverAddress)
+
+	// Step 1 - Verify the transfer of Hbars to the Bridge Account
```
Comment (Daniel-K-Ivanov) [inline e2e/e2e_test.go:107]: nit: Shouldn't be `Hbars` but `HTS Token`

File: e2e/e2e_test.go [hunk 2]
```diff
@@ -180,6 +198,59 @@ func verifyTransferToBridgeAccount(setup *setup.Setup, memo string, whbarReceive
...
+	}
+	transactionReceipt, err := transactionResponse.GetReceipt(setup.Clients.Hedera)
+	if err != nil {
+		fmt.Println(fmt.Sprintf(`Transaction unsuccessful, Error: [%s]`, err))
+		t.Fatal(err)
+	}
+
+	fmt.Println(fmt.Sprintf(`Successfully sent Tokens to bridge account, Status: [%s]`, transactionReceipt.Status))
+
+	// Get bridge account hbar balance after transfer
```
Comment (Daniel-K-Ivanov) [inline e2e/e2e_test.go:233]: shouldn't be `hbar` but `HTS token`

File: e2e/e2e_test.go [hunk 3]
```diff
@@ -180,6 +198,59 @@ func verifyTransferToBridgeAccount(setup *setup.Setup, memo string, whbarReceive
...
+	fmt.Println(fmt.Sprintf(`Bridge Account Token balance after transaction: [%d]`, receiverBalanceNew.Token[setup.TokenID]))
+
+	// Verify that the custodial address has received exactly the amount sent
+	amount := receiverBalanceNew.Token[setup.TokenID] - receiverBalance.Token[setup.TokenID]
+	// Verify that the bridge account has received exactly the amount sent
+	if amount != uint64(tokensSendAmount) {
+		t.Fatalf(`Expected to recieve the exact transfer amount of hbar: [%v]`, hBarSendAmount.AsTinybar())
+	}
+
+	return *transactionResponse, whbarBalanceBefore
```
Comment (Daniel-K-Ivanov) [inline e2e/e2e_test.go:251]: if we are not using `whbarBalanceBefore` why don't you remove it?

File: e2e/e2e_test.go [hunk 4]
```diff
@@ -180,6 +198,59 @@ func verifyTransferToBridgeAccount(setup *setup.Setup, memo string, whbarReceive
 	return *transactionResponse, whbarBalanceBefore
 }
 
+func verifyTokenTransferToBridgeAccount(setup *setup.Setup, memo string, wTokenReceiverAddress common.Address, t *testing.T) (hedera.TransactionResponse, *big.Int) {
+	// Get the wrapped hbar balance of the receiver before the transfer
+	whbarBalanceBefore, err := setup.Clients.WTokenContract.BalanceOf(&bind.CallOpts{}, wTokenReceiverAddress)
```
Comment (Daniel-K-Ivanov) [inline e2e/e2e_test.go:203]: this is not whbar. Its `wrappedToken`

File: e2e/e2e_test.go [hunk 5]
```diff
@@ -250,7 +343,7 @@ func verifyTopicMessages(setup *setup.Setup, transactionResponse hedera.Transact
 	}
 
 	select {
-	case <-time.After(60 * time.Second):
+	case <-time.After(120 * time.Second):
```
Comment (Daniel-K-Ivanov) [inline e2e/e2e_test.go:346]: if there a reason for the increased time here?

File: e2e/e2e_test.go [hunk 6]
```diff
@@ -180,6 +198,59 @@ func verifyTransferToBridgeAccount(setup *setup.Setup, memo string, whbarReceive
...
+	amount := receiverBalanceNew.Token[setup.TokenID] - receiverBalance.Token[setup.TokenID]
+	// Verify that the bridge account has received exactly the amount sent
+	if amount != uint64(tokensSendAmount) {
+		t.Fatalf(`Expected to recieve the exact transfer amount of hbar: [%v]`, hBarSendAmount.AsTinybar())
+	}
+
+	return *transactionResponse, whbarBalanceBefore
+}
+
 func sendHbarsToBridgeAccount(setup *setup.Setup, memo string) (*hedera.TransactionResponse, error) {
```
Comment (Daniel-K-Ivanov) [inline e2e/e2e_test.go:254]: `sendHbarsToBridgeAccount` and `sendTokensToBridgeAccount` can be moved in `./clients/hedera.go` file. We will reduce the bloat in the main `e2e_test.go` file that way

File: e2e/e2e_test.go [hunk 7]
```diff
@@ -94,6 +96,22 @@ func Test_E2E_Only_Address_Memo(t *testing.T) {
...
+	setupEnv := setup.Load()
+
+	memo := fmt.Sprintf("%s-0-0", receiverAddress)
+
+	// Step 1 - Verify the transfer of Hbars to the Bridge Account
+	transactionResponse, _ := verifyTokenTransferToBridgeAccount(setupEnv, memo, whbarReceiverAddress, t)
+
+	// Step 2 - Verify the submitted topic messages
+	verifyTopicMessages(setupEnv, transactionResponse, 0, t)
+	//TODO: submit the tx
```
Comment (Daniel-K-Ivanov) [inline e2e/e2e_test.go:112]: You have to add a new function here: `verifyTransferFromAPI` that would:
1. Query the validator (add new function in the `./clients/validator.go` for querying the transfer data)
2. Verify the returned data

After that you can add the next step: `SubmitAndVerifyMint`, which would use the returned data from the validator API and call the contract mint function. 
Wait for the transaction to be successfully mined.
Verify the token balance of the receiver

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]

File: e2e/e2e_test.go [hunk 8]
```diff
@@ -113,9 +118,9 @@ func Test_No_Ethereum_TX_Submission(t *testing.T) {
 
 	memo := fmt.Sprintf("%s-0-0", receiverAddress)
 
-	// Step 1 - Verify the transfer of Hbars to the Bridge Account
-	transactionResponse, _ := verifyTransferToBridgeAccount(setupEnv, memo, whbarReceiverAddress, t)
-
+	// Step 1 - Verify the transfer of HTS Token to the Bridge Account
```
Comment (Daniel-K-Ivanov) [inline e2e/e2e_test.go:121]: This test should be left to work with hbars. We can change the name to `Test_HBAR_No_Ethereum_TX_Submission` as-well

File: e2e/e2e_test.go [hunk 9]
```diff
@@ -113,9 +118,9 @@ func Test_No_Ethereum_TX_Submission(t *testing.T) {
...
+	transactionResponse, _ := verifyTokenTransferToBridgeAccount(setupEnv, memo, whbarReceiverAddress, t)
+	fmt.Println("transactionResponse:", transactionResponse.TransactionID)
```
Comment (Daniel-K-Ivanov) [inline e2e/e2e_test.go:123]: this line is unnecessary maybe?

File: e2e/e2e_test.go [hunk 10]
```diff
@@ -136,6 +141,116 @@ func Test_No_Ethereum_TX_Submission(t *testing.T) {
...
 }
 
+func Test_E2E_Token_Transfer(t *testing.T) {
+	setupEnv := setup.Load()
+
+	memo := fmt.Sprintf("%s-0-0", receiverAddress)
+
+	wTokenReceiverAddress := common.HexToAddress(receiverAddress)
+
+	// Step 1 - Verify the transfer of Hbars to the Bridge Account
```
Comment (Daniel-K-Ivanov) [inline e2e/e2e_test.go:151]: Should be `HTS` instead of `Hbars`

File: e2e/e2e_test.go [hunk 11]
```diff
@@ -136,6 +141,116 @@ func Test_No_Ethereum_TX_Submission(t *testing.T) {
...
+	transactionResponse, wrappedTokenBalanceBefore := verifyTokenTransferToBridgeAccount(setupEnv, memo, wTokenReceiverAddress, t)
+
+	// Step 2 - Verify the submitted topic messages
+	_, receivedSignatures := verifyTopicMessages(setupEnv, transactionResponse, 0, t)
+
+	// tokenAddress, _ := setup.ParseToken(setupEnv.Clients.RouterContract, setupEnv.TokenID.String())
```
Comment (Daniel-K-Ivanov) [inline e2e/e2e_test.go:157]: I guess this is unnecessary

File: e2e/e2e_test.go [hunk 12]
```diff
@@ -257,6 +372,58 @@ func verifyTransferToBridgeAccount(setup *setup.Setup, memo string, whbarReceive
 	return *transactionResponse, whbarBalanceBefore
 }
 
+func verifyTokenTransferToBridgeAccount(setup *setup.Setup, memo string, wTokenReceiverAddress common.Address, t *testing.T) (hedera.TransactionResponse, *big.Int) {
+	// Get the wrapped hbar balance of the receiver before the transfer
```
Comment (Daniel-K-Ivanov) [inline e2e/e2e_test.go:376]: `hbar` -> `hts token`

File: e2e/e2e_test.go [hunk 13]
```diff
@@ -257,6 +372,58 @@ func verifyTransferToBridgeAccount(setup *setup.Setup, memo string, whbarReceive
...
+	wrappedTokenBalanceBefore, err := setup.Clients.WTokenContract.BalanceOf(&bind.CallOpts{}, wTokenReceiverAddress)
+	if err != nil {
+		log.Fatal(err)
```
Comment (Daniel-K-Ivanov) [inline e2e/e2e_test.go:379]: should be `t.Fatal` i guess with a log

File: e2e/e2e_test.go [hunk 14]
```diff
@@ -136,6 +141,116 @@ func Test_No_Ethereum_TX_Submission(t *testing.T) {
...
+	c1 := make(chan bool, 1)
+	onSuccess := func() {
+		fmt.Printf("Transaction [%s] mined successfully\n", txHash)
+		c1 <- true
+	}
+	onRevert := func() {
+		t.Fatalf(`Failed to mine successfully ethereum transaction: [%s]`, txHash)
+	}
+	onError := func(err error) {
+		if err != nil {
```
Comment (Daniel-K-Ivanov) [inline e2e/e2e_test.go:201]: I think that we do not need the `if` here

File: e2e/e2e_test.go [hunk 15]
```diff
@@ -136,6 +141,116 @@ func Test_No_Ethereum_TX_Submission(t *testing.T) {
...
+
+	// tokenAddress, _ := setup.ParseToken(setupEnv.Clients.RouterContract, setupEnv.TokenID.String())
+
+	transactionData, tokenAddress := verifyTransactionData(setupEnv, transactionResponse, receivedSignatures, t)
+
+	txHash := executeTransaction(setupEnv, transactionResponse, transactionData, tokenAddress, t)
+
+	validateTransactionReceipt(setupEnv, txHash, t)
+
+	validateTokenBalance(setupEnv, wrappedTokenBalanceBefore, wTokenReceiverAddress, t)
```
Comment (Daniel-K-Ivanov) [inline e2e/e2e_test.go:165]: Lets pack them into steps:
```
// Step 3 - Verify Transfer retrieved from Validator API
transactionData, tokenAddress := verifyTransferFromValidatorAPI(setupEnv, transactionResponse, receivedSignatures, t)

// Step 4 - Submit Mint transaction
txHash := submitMintTransaction(setupEnv, transactionResponse, transactionData, tokenAddress, t)

// Step 5 - Wait for transaction to be mined
waitForTransaction(setupEnv, txHash, t)

// Step 6 - Validate Token balances
validateTokenBalance(setupEnv, wrappedTokenBalanceBefore, wTokenReceiverAddress, t)

File: e2e/e2e_test.go [hunk 16]
```diff
@@ -136,6 +141,116 @@ func Test_No_Ethereum_TX_Submission(t *testing.T) {
...
+}
```
Comment (Daniel-K-Ivanov) [inline e2e/e2e_test.go:166]: We should add one more step to verify DB records after execution. See the e2e test above.

```Go
// Step 7 - Verify Database records
expectedTxRecord := prepareExpectedTransfer(
		setupEnv.Clients.RouterContract,
		transactionResponse.TransactionID,
		nativeToken,
		"0",
		"0",
		database.ExpectedStatuses{
			Status:          entity_transfer.StatusCompleted,
			StatusSignature: entity_transfer.StatusSignatureMined,
		},
		ethTransactionHash,
		false, t)
	verifyDatabaseRecords(setupEnv.DbValidation, expectedTxRecord, receivedSignatures, t)
```

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]

**Comment by Daniel-K-Ivanov** [comment]
Lets resolve the E2E test before merging

**Review by Daniel-K-Ivanov — APPROVED** [review]

File: e2e/e2e_test.go [hunk 17]
```diff
@@ -136,6 +141,134 @@ func Test_No_Ethereum_TX_Submission(t *testing.T) {
...
+	<-c1
+}
+
+func validateTokenBalance(setupEnv *setup.Setup, wrappedTokenBalanceBefore *big.Int, wTokenReceiverAddress common.Address, t *testing.T) {
+	wrappedTokenBalanceAfter, err := setupEnv.Clients.WTokenContract.BalanceOf(&bind.CallOpts{}, wTokenReceiverAddress)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tokensAmount := new(big.Int).SetInt64(int64(tokensSendAmount))
```
Comment (georgiyazovaliiski) [inline e2e/e2e_test.go:231]: I think `big.NewInt(tokensSendAmount)` will do what you are looking for here.

File: e2e/e2e_test.go [hunk 18]
```diff
@@ -136,6 +141,134 @@ func Test_No_Ethereum_TX_Submission(t *testing.T) {
...
+	}
+
+	txFee := new(big.Int).Mul(tokensAmount, serviceFeePercentage)
+	txFee = new(big.Int).Div(txFee, precision)
+
+	expectedBalance := new(big.Int).Sub(wrappedTokenBalanceAfter, wrappedTokenBalanceBefore)
+	mintAmount := new(big.Int).Sub(tokensAmount, txFee)
+
+	if expectedBalance.Cmp(mintAmount) != 0 {
+		t.Fatalf("Incorect token balance")
```
Comment (georgiyazovaliiski) [inline e2e/e2e_test.go:245]: `Incorrect token balance. Expected to be [%s], but was [%s].`

File: e2e/e2e_test.go [hunk 19]
```diff
@@ -136,6 +141,134 @@ func Test_No_Ethereum_TX_Submission(t *testing.T) {
...
+	}
+	if transactionData.NativeToken != setupEnv.TokenID.String() {
+		t.Fatal("Native Token mismatch")
+	}
+	if transactionData.Recipient != receiverAddress {
+		t.Fatal("Receiver address mismatch")
+	}
+	if transactionData.WrappedToken != tokenAddress.String() {
+		t.Fatal("Token address mismatch")
+	}
```
Comment (georgiyazovaliiski) [inline e2e/e2e_test.go:267]: Can we make the fatal logs a bit more descriptive, similar to this suggestion?
`Incorrect token balance. Expected to be [%s], but was [%s].`

File: e2e/e2e_test.go [hunk 20]
```diff
@@ -257,6 +390,59 @@ func verifyTransferToBridgeAccount(setup *setup.Setup, memo string, whbarReceive
...
 
+func verifyTokenTransferToBridgeAccount(setup *setup.Setup, memo string, wTokenReceiverAddress common.Address, t *testing.T) (hedera.TransactionResponse, *big.Int) {
+	// Get the wrapped hts token balance of the receiver before the transfer
+	wrappedTokenBalanceBefore, err := setup.Clients.WTokenContract.BalanceOf(&bind.CallOpts{}, wTokenReceiverAddress)
+	if err != nil {
+		fmt.Println(`Unable to query the token balance of the  receiver account`)
+		t.Fatal(err)
+	}
+
+	fmt.Printf("Token balance before transaction: [%s]\n", wrappedTokenBalanceBefore)
```
Comment (georgiyazovaliiski) [inline e2e/e2e_test.go:401]: `fmt.Println(fmt.Sprintf())`

File: e2e/e2e_test.go [hunk 21]
```diff
@@ -257,6 +390,59 @@ func verifyTransferToBridgeAccount(setup *setup.Setup, memo string, whbarReceive
...
+		fmt.Println(`Unable to query the token balance of the Bridge Account`)
+		t.Fatal(err)
+	}
+	fmt.Println(fmt.Sprintf(`Bridge account Token balance before transaction: [%d]`, receiverBalance.Token[setup.TokenID]))
+
+	// Get the transaction receipt to verify the transaction was executed
+	transactionResponse, err := sendTokensToBridgeAccount(setup, memo)
+	if err != nil {
+		fmt.Println(fmt.Sprintf(`Unable to send Tokens to Bridge Account, Error: [%s]`, err))
+		t.Fatal(err)
```
Comment (georgiyazovaliiski) [inline e2e/e2e_test.go:416]: We can merge these two lines into one: `t.Fatalf(`Unable to send Tokens to Bridge Account, Error: [%s]`, err)`

File: e2e/e2e_test.go [hunk 22]
```diff
@@ -257,6 +390,59 @@ func verifyTransferToBridgeAccount(setup *setup.Setup, memo string, whbarReceive
...
+	}
+	transactionReceipt, err := transactionResponse.GetReceipt(setup.Clients.Hedera)
+	if err != nil {
+		fmt.Println(fmt.Sprintf(`Transaction unsuccessful, Error: [%s]`, err))
+		t.Fatal(err)
```
Comment (georgiyazovaliiski) [inline e2e/e2e_test.go:421]: Can we merge these two?

File: e2e/e2e_test.go [hunk 23]
```diff
@@ -257,6 +390,59 @@ func verifyTransferToBridgeAccount(setup *setup.Setup, memo string, whbarReceive
...
+	}
+	fmt.Println(fmt.Sprintf(`Successfully sent Tokens to bridge account, Status: [%s]`, transactionReceipt.Status))
+
+	// Get bridge account HTS token balance after transfer
+	receiverBalanceNew, err := hedera.NewAccountBalanceQuery().
+		SetAccountID(setup.BridgeAccount).
+		Execute(setup.Clients.Hedera)
+	if err != nil {
+		fmt.Println(`Unable to query the token balance of the Bridge Account`)
+		t.Fatal(err)
```
Comment (georgiyazovaliiski) [inline e2e/e2e_test.go:431]: Merge?

File: e2e/e2e_test.go [hunk 24]
```diff
@@ -257,6 +390,59 @@ func verifyTransferToBridgeAccount(setup *setup.Setup, memo string, whbarReceive
...
+	}
+
+	fmt.Println(fmt.Sprintf(`Bridge Account Token balance after transaction: [%d]`, receiverBalanceNew.Token[setup.TokenID]))
+
+	// Verify that the custodial address has received exactly the amount sent
+	amount := receiverBalanceNew.Token[setup.TokenID] - receiverBalance.Token[setup.TokenID]
+	// Verify that the bridge account has received exactly the amount sent
+	if amount != uint64(tokensSendAmount) {
+		t.Fatalf(`Expected to recieve the exact transfer amount of hbar: [%v]`, hBarSendAmount.AsTinybar())
```
Comment (georgiyazovaliiski) [inline e2e/e2e_test.go:440]: recieve -> receive

Comment (georgiyazovaliiski) [reply e2e/e2e_test.go:440]: can we add, `but was [%v]`

File: e2e/setup/config.go [hunk 25]
```diff
@@ -143,13 +154,17 @@ func newClients(config Config) (*clients, error) {
 
 	validatorClient := e2eClients.NewValidatorClient(config.ValidatorUrl)
 
+	signer := eth.NewEthSigner(config.Signer)
+	keyTransactor, _ := signer.NewKeyTransactor(ethClient.ChainID())
```
Comment (georgiyazovaliiski) [inline e2e/setup/config.go:158]:
```
if err != nil {
	return nil, err
}
```

**Review by georgiyazovaliiski — CHANGES_REQUESTED** [review]

**Review by georgiyazovaliiski — APPROVED** [review]

**Review by Daniel-K-Ivanov — APPROVED** [review]

**Pull request merged**